bigquery/tests/unit/test_table.py

```diff
@@ -871,6 +871,54 @@ def test__build_resource_w_custom_field_not_in__properties(self):
         with self.assertRaises(ValueError):
             table._build_resource(["bad"])
 
+    def test_time_partitioning_getter(self):
+        from google.cloud.bigquery.table import TimePartitioning
+        from google.cloud.bigquery.table import TimePartitioningType
+
+        dataset = DatasetReference(self.PROJECT, self.DS_ID)
+        table_ref = dataset.table(self.TABLE_NAME)
+        table = self._make_one(table_ref)
+
+        table._properties["timePartitioning"] = {
+            "type": "DAY",
+            "field": "col1",
+            "expirationMs": "123456",
+            "requirePartitionFilter": False,
+        }
+        self.assertIsInstance(table.time_partitioning, TimePartitioning)
+        self.assertEqual(table.time_partitioning.type_, TimePartitioningType.DAY)
+        self.assertEqual(table.time_partitioning.field, "col1")
+        self.assertEqual(table.time_partitioning.expiration_ms, 123456)
+        self.assertFalse(table.time_partitioning.require_partition_filter)
+
+    def test_time_partitioning_getter_w_none(self):
+        dataset = DatasetReference(self.PROJECT, self.DS_ID)
+        table_ref = dataset.table(self.TABLE_NAME)
+        table = self._make_one(table_ref)
+
+        table._properties["timePartitioning"] = None
+        self.assertIsNone(table.time_partitioning)
+
+        del table._properties["timePartitioning"]
+        self.assertIsNone(table.time_partitioning)
+
+    def test_time_partitioning_getter_w_empty(self):
+        from google.cloud.bigquery.table import TimePartitioning
+
+        dataset = DatasetReference(self.PROJECT, self.DS_ID)
+        table_ref = dataset.table(self.TABLE_NAME)
+        table = self._make_one(table_ref)
+
+        # Even though there are required properties according to the API
+        # specification, sometimes time partitioning is populated as an empty
+        # object. See internal bug 131167013.
+        table._properties["timePartitioning"] = {}
+        self.assertIsInstance(table.time_partitioning, TimePartitioning)
+        self.assertIsNone(table.time_partitioning.type_)
+        self.assertIsNone(table.time_partitioning.field)
+        self.assertIsNone(table.time_partitioning.expiration_ms)
+        self.assertIsNone(table.time_partitioning.require_partition_filter)
+
```

Comment:
We need a test for `TimePartitioning.from_api_repr` that exercises the bug I corrected above (stomping on the type from the resource).

Reply:
Good call, done in 14a0495.